### PR TITLE
feat: allow users to bypass host preflights in install wizard  

### PR DESCRIPTION
**Pull request by diamonwiggins**
#### What this PR does / why we need it:

This PR implements the `--ignore-host-preflights` CLI flag functionality in the web installation wizard, allowing users who set the flag to optionally proceed with installation when host preflights fail. The implementation includes a dual validation security model requiring both the CLI flag and explicit user confirmation.

**Key Changes**:
- **Backend**: Added `AllowIgnoreHostPreflights` field to API response types and implemented comprehensive validation logic throughout the install controller
- **Frontend**: Created reusable `Modal` component and implemented conditional UI logic in `ValidationStep` and `LinuxPreflightCheck` components
- **API Integration**: Extended preflight status endpoints to include bypass permission flag
- **Security**: Implemented dual validation requiring both CLI flag presence and user confirmation
- **Testing**: Comprehensive test coverage with 10+ new tests across frontend and backend components

**Technical Implementation**:
- Backend API properly propagates CLI flag state through `AllowIgnoreHostPreflights` field
- Frontend conditionally enables/disables "Start Installation" button based on API response
- Warning modal appears when user attempts to proceed with failed preflights (only when CLI flag is set)
- Integration tests validate end-to-end flow including API parameter passing

#### Which issue(s) this PR fixes:

[SC-124346](https://app.shortcut.com/replicated/story/124346/a-user-can-proceed-the-installation-even-when-there-are-preflight-failures)

#### Does this PR require a test?

**YES** - Comprehensive test coverage implemented:
- **Frontend Tests**: Modal component tests, ValidationStep behavior tests, LinuxPreflightCheck integration tests
- **Backend Tests**: Install controller tests, API integration tests, host preflights endpoint tests  
- **Integration Tests**: End-to-end validation of CLI flag → API → Frontend flow
- **Edge Cases**: Tests for various combinations of CLI flag states and preflight results

#### Does this PR require a release note?

```release-note
NONE
```

#### Does this PR require documentation?

NONE

### CODE REVIEW

File: web/src/components/wizard/ValidationStep.tsx [hunk 1]
```diff
@@ -66,13 +122,47 @@ const ValidationStep: React.FC<ValidationStepProps> = ({ onNext, onBack }) => {
...
+        }
+      >
+        <div className="flex items-start space-x-3">
+          <div className="flex-shrink-0">
+            <AlertTriangle className="h-6 w-6 text-amber-500" />
+          </div>
+          <div>
+            <p className="text-sm text-gray-700">
+              Some validation checks have failed. Are you sure you want to continue with the installation? 
+              This may cause installation issues.
```
Comment (ajp-io) [inline web/src/components/wizard/ValidationStep.tsx:161]:
```suggestion
              Some host preflight checks have failed. Continuing with the installation is likely to cause errors. Are you sure you want to proceed?
```

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: api/integration/infra_test.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+	"github.com/replicatedhq/embedded-cluster/api"
+	"github.com/replicatedhq/embedded-cluster/api/controllers/install"
+	"github.com/replicatedhq/embedded-cluster/api/pkg/logger"
+	"github.com/replicatedhq/embedded-cluster/api/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// Simple mock controller for infra setup validation tests
+type mockInfraController struct {
```
[the remark this answers is not in the document]
Comment (diamonwiggins) [reply api/integration/infra_test.go:22]: yeah, i wasn't clear if this was the right thing to do either

File: api/controllers/install/infra.go [hunk 3]
```diff
@@ -2,11 +2,14 @@ package install
 
 import (
 	"context"
+	"errors"
 	"fmt"
 
 	"github.com/replicatedhq/embedded-cluster/api/types"
 )
 
+var ErrPreflightChecksFailed = errors.New("Preflight checks failed")
```
Comment (emosbaugh) [inline api/controllers/install/infra.go:11]: go error strings should be lowercase

File: api/controllers/install/infra.go [hunk 4]
```diff
@@ -22,23 +25,25 @@ func (c *InstallController) SetupInfra(ctx context.Context, ignorePreflightFailu
...
+			return false, fmt.Errorf("get install host preflight output: %w", err)
 		}
 
 		// Check if we can proceed despite failures
 		if !ignorePreflightFailures || !c.allowIgnoreHostPreflights {
-			return false, fmt.Errorf("Preflight checks failed")
+			// Report metrics for failed preflights that are NOT bypassed
+			if c.metricsReporter != nil && preflightOutput != nil {
+				c.metricsReporter.ReportPreflightsFailed(ctx, preflightOutput)
+			}
```
Comment (emosbaugh) [inline api/controllers/install/infra.go:39]: This is not needed as it should have [already](https://github.com/replicatedhq/embedded-cluster/blob/5c52173fa644edea3aa6dc4026063d5774d99e4e/api/internal/managers/preflight/hostpreflight.go#L141) been reported when they failed.

**Comment by emosbaugh** [comment]
Looks good. Please hold off on merging this

File: web/src/components/common/Modal.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+interface ModalProps {
+  isOpen: boolean;
+  onClose: () => void;
+  title: string;
+  children: React.ReactNode;
+  footer?: React.ReactNode;
+}
+
+export const Modal: React.FC<ModalProps> = ({ isOpen, onClose, title, children, footer }) => {
+  if (!isOpen) return null;
```
Comment (JGAntunes) [inline web/src/components/common/Modal.tsx:13]: As it stands, won't this render a `null` in our DOM if `isOpen` is false?

I wonder if the decision to render this should be at the top level component instead, similar to things like:
- https://github.com/replicatedhq/embedded-cluster/blob/5c52173fa644edea3aa6dc4026063d5774d99e4e/web/src/components/wizard/InstallationStep.tsx#L94-L95
- https://github.com/replicatedhq/embedded-cluster/blob/29678df88daa873458a08e7394f0b434a1951ee9/web/src/components/wizard/installation/LogViewer.tsx#L39-L51

File: api/controllers/install/infra.go [hunk 6]
```diff
@@ -2,36 +2,53 @@ package install
...
+	// Check preflight status and requirements
 	preflightStatus, err := c.GetHostPreflightStatus(ctx)
 	if err != nil {
-		return fmt.Errorf("get install host preflight status: %w", err)
+		return false, fmt.Errorf("get install host preflight status: %w", err)
 	}
 
 	if preflightStatus.State != types.StateFailed && preflightStatus.State != types.StateSucceeded {
-		return fmt.Errorf("host preflight checks did not complete")
+		return false, fmt.Errorf("host preflight checks did not complete")
```
[the remark this answers is not in the document]
Comment (JGAntunes) [reply api/controllers/install/infra.go:21]: Wrote this before seeing the error handler at the HTTP handler, see comment for further discussion:
- https://github.com/replicatedhq/embedded-cluster/pull/2335#discussion_r2158419960

File: api/controllers/install/infra.go [hunk 7]
```diff
@@ -2,36 +2,53 @@ package install
...
 		if err != nil {
-			return fmt.Errorf("get install host preflight output: %w", err)
+			return false, fmt.Errorf("get install host preflight output: %w", err)
 		}
-		if preflightOutput != nil {
-			c.metricsReporter.ReportPreflightsFailed(ctx, preflightOutput)
+
+		// Check if we can proceed despite failures
+		if !ignorePreflightFailures || !c.allowIgnoreHostPreflights {
+			return false, ErrPreflightChecksFailed
```
[the remark this answers is not in the document]
Comment (JGAntunes) [reply api/controllers/install/infra.go:36]: Wrote this before seeing the error handler at the HTTP handler, see comment for further discussion:
- https://github.com/replicatedhq/embedded-cluster/pull/2335#discussion_r2158419960

File: api/design/infra-setup-preflight-validation.md [hunk 8]
```diff
@@ -0,0 +1,238 @@
+# Infrastructure Setup API with Preflight Validation
```
[the remark this answers is not in the document]
Comment (diamonwiggins) [reply api/design/infra-setup-preflight-validation.md:1]: I'd like to see more of this in the repo personally, but i'm also coming at it from the perspective of someone who uses AI quite heavily. Keeping it updated isn't a ton of work, but remembering to do so I think would be more of the problem.

[the remark this answers is not in the document]
Comment (diamonwiggins) [reply api/design/infra-setup-preflight-validation.md:1]: I may decide to scrap it in the near term given the amount of foundational changes happening to the API at the moment.

File: web/src/components/common/Modal.test.tsx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+            <p>Modal content</p>
+          </Modal>
+        )}
+      </div>
+    );
+    
+    expect(screen.getByText('Test Modal')).toBeInTheDocument();
+    expect(screen.getByText('Modal content')).toBeInTheDocument();
+  });
+
```
[the remark this answers is not in the document]
Comment (diamonwiggins) [reply web/src/components/common/Modal.test.tsx:49]: agreed, going to remove

File: api/integration/install_test.go [hunk 10]
```diff
@@ -109,8 +121,28 @@ func (m *mockInstallController) GetHostPreflightTitles(ctx context.Context) ([]s
 	return []string{}, nil
 }
 
-func (m *mockInstallController) SetupInfra(ctx context.Context) error {
-	return m.setupInfraError
+func (m *mockInstallController) SetupInfra(ctx context.Context, ignorePreflightFailures bool) error {
```
[the remark this answers is not in the document]
Comment (diamonwiggins) [reply api/integration/install_test.go:111]: Should i make this change on this PR or something to follow up with?

File: api/controllers/install/infra.go [hunk 11]
```diff
@@ -2,18 +2,34 @@ package install
...
+var (
+	ErrPreflightChecksFailed      = errors.New("preflight checks failed")
+	ErrPreflightChecksNotComplete = errors.New("preflight checks not complete")
+)
+
+func (c *InstallController) SetupInfra(ctx context.Context, ignorePreflightFailures bool) (finalErr error) {
+	currentState := c.stateMachine.CurrentState()
+
+	// Check if preflights are complete
+	switch currentState {
```
Comment (sgalsaleh) [inline api/controllers/install/infra.go:21]: why is this switch statement necessary? it seems like something we should leave to the state machine to return conflict errors based on state, and should simply pass the `ignorePreflightFailures` bool to the `c.bypassPreflights` function.

File: api/install.go [hunk 12]
```diff
@@ -148,18 +149,35 @@ func (a *API) getInstallHostPreflightsStatus(w http.ResponseWriter, r *http.Requ
...
+
+	// Setup infrastructure with preflight validation handled internally
+	err := a.installController.SetupInfra(r.Context(), req.IgnorePreflightFailures)
 	if err != nil {
 		a.logError(r, err, "failed to setup infra")
 		a.jsonError(w, r, err)
 		return
 	}
 
-	a.getInstallInfraStatus(w, r)
```
Comment (sgalsaleh) [inline api/install.go:162]: why is this change necessary? seems redundant

File: api/types/infra.go [hunk 13]
```diff
@@ -1,5 +1,10 @@
 package types
 
+// InfraSetupRequest represents a request to set up infrastructure
+type InfraSetupRequest struct {
+	IgnorePreflightFailures bool `json:"ignorePreflightFailures"`
```
Comment (sgalsaleh) [inline api/types/infra.go:5]: the cli flag and variable name we pass to the api is named `allowIgnoreHostPreflights`, i think we should keep naming consistent and name this `ignoreHostPreflights`.

**Review by sgalsaleh — APPROVED** [review]
